w3ctrace/state.go

```diff
@@ -11,24 +11,56 @@ import (
 // VendorInstana is the Instana vendor key in the `tracestate` list
 const VendorInstana = "in"
 
+// Max amount of KV pairs in `tracestate` header
+const maxKVPairs = 32
+
+// Length of entries that should be filtered first in case, if tracestate has more than `maxKVPairs` items
+const thresholdLen = 128
+
 // State is list of key=value pairs representing vendor-specific data in the trace context
 type State []string
 
-// ParseState parses the value of `tracestate` header. It strips any optional white-space chararacters
-// preceding or following the key=value pairs. Empty list items are omitted.
+// ParseState parses the value of `tracestate` header. Empty list items are omitted.
 func ParseState(traceStateValue string) (State, error) {
 	var state State
 
-	for _, st := range strings.SplitN(traceStateValue, ",", 32) {
-		st = strings.TrimSpace(st)
-		if st == "" {
-			continue
+	entries := strings.Split(traceStateValue, ",")
+	if len(entries) == 1 && entries[0] == "" {
+		return state, nil
+	}
+
+	filteredEntries := filterEmptyItems(entries)
+	filteredEntriesLen := len(filteredEntries)
+
+	if filteredEntriesLen == 0 {
+		return state, nil
+	}
+
+	if filteredEntriesLen > maxKVPairs {
+		filtered := 0
+
+		for k, st := range filteredEntries {
+			if len(state) == maxKVPairs {
+				break
+			}
+
+			// check if enough elements were filtered already
+			if filteredEntriesLen-filtered <= maxKVPairs {
+				return append(state, filteredEntries[k:]...), nil
+			}
```

Comment:
I still find it hard to understand the logic here, since one needs to consider the size two lists at once. I believe we can do this with one list, removing the items longer than 128 chars until either it contains 32 elements, or there are no large elements left. And then return first `min(len(state), 32)` elements of it?

Reply:
Thank you for the idea. I have changed the implementation and added a new test case.

1. I removed extra memory allocations for filtering.
2. I changed the filtering function, now I think it looks close to what you propose.

Please check when you will have time.